### PR TITLE
Reset node-left generation on election

### PR DESCRIPTION
**Pull request by DaveCTurner**
Node-left generations are only comparable within a master term and must
start again from zero in the next term.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java [hunk 1]
```diff
@@ -855,7 +856,10 @@ public DiscoveryNodes build() {
             } else if (this.nodeLeftGeneration != null) {
                 // only happens during deserialization
                 assert removedNode == false;
+                assert resetNodeLeftGeneration == false;
                 newNodeLeftGeneration = nodeLeftGeneration;
+            } else if (resetNodeLeftGeneration) {
+                newNodeLeftGeneration = 0L;
```
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:862]: Should we also set `resetNodeLeftGeneration` to false? Or it does not matter?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:862]: We basically never re-use builders after calling `.build()` so it won't really matter, but if we did re-use the builder I think we would want to preserve the `resetNodeLeftGeneration` flag.

**Review by fcofdez — APPROVED** [review]
LGTM

**Review by kingherc — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
I'll revert that change after this has landed in the serverless submodule, otherwise we'll just see spurious failures I think.

**Pull request merged**